Tests/test_permutation.py

```
@@ -0,0 +1,87 @@
+    ("5", 2),
+    (5, "2"),
+    (None, 2),
+])
+def test_n_permute_r_invalid_types(n, r):
+    """Test n_permute_r raises TypeError for non-integer inputs."""
+    with pytest.raises(TypeError):
+        n_permute_r(n, r)
+
+def test_n_permute_r_value_error_message():
+    """Test that n_permute_r raises ValueError with the correct message."""
+    with pytest.raises(ValueError, match="n should be greater than or equal to r for permutations to be valid."):
```

Ruff (0.16.0)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)